opt: Add some scalar normalization rules

andy-kimball opened the pull request:
Add the set of scalar normalization rules that are currently
performed by opt/scalar.go (plus a couple):
  EliminateAnd: eliminate And if it has a False condition
  EliminateOr: eliminate Or if it has a True condition
  FlattenAnd: flatten nested And ops into one list
  FlattenOr: flatten nested Or ops into one list
  NegateComparison: flip cases like Not(x=1) to x<>1
  EliminateNot: eliminate double Not(Not(..))
  NormalizeVar: ensure vars are on left side of Eq/Ne
  NormalizeTupleEquality: break (a, b) = (x, y) into (a=x) AND (b=y)

This prepares for the next step, which is to convert the index
constraints code to use the new exprs. It also generates rules for
the first time, which surfaced a small bug in Optgen where it was
incorrectly capitalizing an expression field name.

Other changes triggered by adding these rules:
  - Remove ContainedBy op, since it's trivial inverse of the Contains
    op, which builder now builds instead.
  - Convert And and Or ops to use a list for their conditions rather
    than being binary ops. This makes matching easier, and also means
    we don't need the Filters op.
  - Map boolean datums to True/False ops, making them easier to
    recognize and construct in rules.
  - Use an enum rather than an int as maxSteps parameter to the
    NewOptimizer function, for better clarity.

Release note: None

andy-kimball committed:
opt: Add some scalar normalization rules

Add the set of scalar normalization rules that are currently
performed by opt/scalar.go (plus a couple):
  EliminateAnd: eliminate And if it has a False condition
  EliminateOr: eliminate Or if it has a True condition
  FlattenAnd: flatten nested And ops into one list
  FlattenOr: flatten nested Or ops into one list
  NegateComparison: flip cases like Not(x=1) to x<>1
  EliminateNot: eliminate double Not(Not(..))
  NormalizeVar: ensure vars are on left side of Eq/Ne
  NormalizeTupleEquality: break (a, b) = (x, y) into (a=x) AND (b=y)

This prepares for the next step, which is to convert the index
constraints code to use the new exprs. It also generates rules for
the first time, which surfaced a small bug in Optgen where it was
incorrectly capitalizing an expression field name.

Other changes triggered by adding these rules:
  - Remove ContainedBy op, since it's trivial inverse of the Contains
    op, which builder now builds instead.
  - Convert And and Or ops to use a list for their conditions rather
    than being binary ops. This makes matching easier, and also means
    we don't need the Filters op.
  - Map boolean datums to True/False ops, making them easier to
    recognize and construct in rules.
  - Use an enum rather than an int as maxSteps parameter to the
    NewOptimizer function, for better clarity.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/22505)
<!-- Reviewable:end -->

RaduBerinde commented:
LGTM

---

Review status: 0 of 25 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/sql/opt/xform/rules_test.go, line 68 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22505#-L4qjreScg5XUe8SHII5:-L4qjreScg5XUe8SHII6:bwbhvo2) ([raw file](https://github.com/cockroachdb/cockroach/blob/210d03c2432310f386fb2cfd6cfa079b8efbb6cc/pkg/sql/opt/xform/rules_test.go#L68)):*
> ```Go
> 	cat := testutils.NewTestCatalog()
> 
> 	// CREATE TABLE a (k INT PRIMARY KEY, i INT, f FLOAT, s STRING, j JSON)
> ```

We should use the `exec-ddl` that Rebecca added instead.

---

*[pkg/sql/opt/xform/rules/comp.opt, line 9 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22505#-L4qikh4whs7VZXZOAmX:-L4qikh4whs7VZXZOAmY:bq1drhp) ([raw file](https://github.com/cockroachdb/cockroach/blob/210d03c2432310f386fb2cfd6cfa079b8efbb6cc/pkg/sql/opt/xform/rules/comp.opt#L9)):*
> ```
> # equality and inequality operators.
> [NormalizeVar, Normalize]
> (Eq | Ne
> ```

Why not all comparison operators? (LT, LE, etc). I think index selection code assumes those are normalized as well.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22505#-:-L4qkFyEhLJ1Wne_P1EN:bb6ywmx)*
<!-- Sent from Reviewable.io -->

rytaft commented:
Reviewed 25 of 25 files at r1.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[pkg/sql/opt/build/builder.go, line 315 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22505#-L4qg-swWVvLSJGSa4fE:-L4qg-swWVvLSJGSa4fF:b26adk1) ([raw file](https://github.com/cockroachdb/cockroach/blob/210d03c2432310f386fb2cfd6cfa079b8efbb6cc/pkg/sql/opt/build/builder.go#L315)):*
> ```Go
> 		right := b.buildScalar(t.TypedRight(), inScope)
> 		conditions := b.factory.InternList([]opt.GroupID{left, right})
> 		return b.factory.ConstructAnd(conditions)
> ```

We can't return from these statements (here and the `OrExpr` case below) since the `if inGroupingContext` block below the switch needs to execute.  (Just assign the result to `out` instead.)

---

*[pkg/sql/opt/xform/factory.go, line 112 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22505#-L4qjKSRCRjqR-nz1HOm:-L4qjKSRCRjqR-nz1HOn:b-sqll09) ([raw file](https://github.com/cockroachdb/cockroach/blob/210d03c2432310f386fb2cfd6cfa079b8efbb6cc/pkg/sql/opt/xform/factory.go#L112)):*
> ```Go
> // children.
> func (f *factory) flattenAnd(conditions opt.ListID) opt.GroupID {
> 	list := make([]opt.GroupID, 0, conditions.Length+1)
> ```

Why the +1 here and below in `flattenOr`?

---

*[pkg/sql/opt/xform/optimizer.go, line 38 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22505#-L4qlkWOq8QW_Looc-aA:-L4qlkWOq8QW_Looc-aB:b-p1qciv) ([raw file](https://github.com/cockroachdb/cockroach/blob/210d03c2432310f386fb2cfd6cfa079b8efbb6cc/pkg/sql/opt/xform/optimizer.go#L38)):*
> ```Go
> 	// until the best plan has been found. There is no limit to the number of
> 	// steps that the optimizer will take to get there.
> 	OptimizeAll = OptimizeSteps(^uint(0) >> 1)
> ```

Maybe define `^uint(0) >> 1` somewhere as `maxInt`?

---

*[pkg/sql/opt/xform/rules/bool.opt, line 51 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22505#-L4qn86HyCCeF56ABE9e:-L4qn86HyCCeF56ABE9f:b4pgjet) ([raw file](https://github.com/cockroachdb/cockroach/blob/210d03c2432310f386fb2cfd6cfa079b8efbb6cc/pkg/sql/opt/xform/rules/bool.opt#L51)):*
> ```
> (InvertComparison $input)
> 
> # EliminateNot discards a doubled Not operator.
> ```

Does the order of these rules matter? Seems like you might want to put `EliminateNot` before `NegateComparison`.

---

*[pkg/sql/opt/xform/rules/comp.opt, line 9 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22505#-L4qikh4whs7VZXZOAmX:-L4qnS5VbzX56c4OD-AU:b-ulpkyw) ([raw file](https://github.com/cockroachdb/cockroach/blob/210d03c2432310f386fb2cfd6cfa079b8efbb6cc/pkg/sql/opt/xform/rules/comp.opt#L9)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Why not all comparison operators? (LT, LE, etc). I think index selection code assumes those are normalized as well.
</blockquote></details>

In opttoy I wrote a function which did that - but for those operators you can't just swap the variables, you also have to change LT to GT, etc. I assume Andy will add that later?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22505)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[pkg/sql/opt/xform/optimizer.go, line 38 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22505#-L4qlkWOq8QW_Looc-aA:-L4qqb63TvvMbhWDr2qi:b-lnwr2j) ([raw file](https://github.com/cockroachdb/cockroach/blob/210d03c2432310f386fb2cfd6cfa079b8efbb6cc/pkg/sql/opt/xform/optimizer.go#L38)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Maybe define `^uint(0) >> 1` somewhere as `maxInt`?
</blockquote></details>

can use `intsets.MaxInt`

---

*[pkg/sql/opt/xform/rules/comp.opt, line 9 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22505#-L4qikh4whs7VZXZOAmX:-L4qqmnE8wQI7gempb3g:bna5sth) ([raw file](https://github.com/cockroachdb/cockroach/blob/210d03c2432310f386fb2cfd6cfa079b8efbb6cc/pkg/sql/opt/xform/rules/comp.opt#L9)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

In opttoy I wrote a function which did that - but for those operators you can't just swap the variables, you also have to change LT to GT, etc. I assume Andy will add that later?
</blockquote></details>

Ah, right, we'd need a separate rule for each one.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22505)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
Review status: 7 of 26 files reviewed at latest revision, 6 unresolved discussions.

---

*[pkg/sql/opt/build/builder.go, line 315 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22505#-L4qg-swWVvLSJGSa4fE:-L4rMWUGUOofdUUH3i6q:b2cihtu) ([raw file](https://github.com/cockroachdb/cockroach/blob/210d03c2432310f386fb2cfd6cfa079b8efbb6cc/pkg/sql/opt/build/builder.go#L315)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

We can't return from these statements (here and the `OrExpr` case below) since the `if inGroupingContext` block below the switch needs to execute.  (Just assign the result to `out` instead.)
</blockquote></details>

Whoops, didn't mean to change that. Fixed. Maybe we're missing test(s) to catch it?

---

*[pkg/sql/opt/xform/factory.go, line 112 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22505#-L4qjKSRCRjqR-nz1HOm:-L4rMs2TvJfYLIhkHca0:b-vmr5w5) ([raw file](https://github.com/cockroachdb/cockroach/blob/210d03c2432310f386fb2cfd6cfa079b8efbb6cc/pkg/sql/opt/xform/factory.go#L112)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Why the +1 here and below in `flattenOr`?
</blockquote></details>

Because if we're flattening an And, there's going to be at least one more expression to add to the final list (and maybe more than one).

---

*[pkg/sql/opt/xform/optimizer.go, line 38 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22505#-L4qlkWOq8QW_Looc-aA:-L4rN-wefPDuLnrDQ5qT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/210d03c2432310f386fb2cfd6cfa079b8efbb6cc/pkg/sql/opt/xform/optimizer.go#L38)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

can use `intsets.MaxInt`
</blockquote></details>

Done.

---

*[pkg/sql/opt/xform/rules_test.go, line 68 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22505#-L4qjreScg5XUe8SHII5:-L4qlfvklqKt_vcflufP:b31nlfo) ([raw file](https://github.com/cockroachdb/cockroach/blob/210d03c2432310f386fb2cfd6cfa079b8efbb6cc/pkg/sql/opt/xform/rules_test.go#L68)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

We should use the `exec-ddl` that Rebecca added instead.
</blockquote></details>

Agreed, converted to use it.

---

*[pkg/sql/opt/xform/rules/bool.opt, line 51 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22505#-L4qn86HyCCeF56ABE9e:-L4rUb7ZUqGV5CHrOdM-:b-nqbs38) ([raw file](https://github.com/cockroachdb/cockroach/blob/210d03c2432310f386fb2cfd6cfa079b8efbb6cc/pkg/sql/opt/xform/rules/bool.opt#L51)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Does the order of these rules matter? Seems like you might want to put `EliminateNot` before `NegateComparison`.
</blockquote></details>

They're run in same order as file. In this case, it doesn't matter what order they're in, though. The rules trigger bottom-up, so `EliminateNot` will never trigger before `NegateComparison`.

---

*[pkg/sql/opt/xform/rules/comp.opt, line 9 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22505#-L4qikh4whs7VZXZOAmX:-L4qlHuy2dqycojfKPcr:b-gv1uu8) ([raw file](https://github.com/cockroachdb/cockroach/blob/210d03c2432310f386fb2cfd6cfa079b8efbb6cc/pkg/sql/opt/xform/rules/comp.opt#L9)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Ah, right, we'd need a separate rule for each one.
</blockquote></details>

Yes, we'll add it later on, for the reasons Rebecca states.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22505)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Re the `rules` path, where do you plan to put exploration xforms? I think we should separate them from normalization rules.

---

Review status: 7 of 26 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22505#-:-L4rXxUdcagfTqcMJptQ:b50cnvj)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
So maybe a `rules/norm` and a `rules/explore` directory? Once we get to that point, I can definitely refactor the directories.

---

Review status: 7 of 26 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22505#-:-L4rYmVs-RCwiZVeb3yF:botnpx6)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Sounds good.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22505#-:-L4r_FOm7ZEb50clWjIu:brxp9sq)*
<!-- Sent from Reviewable.io -->

rytaft commented:
Reviewed 19 of 19 files at r2.
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt/build/builder.go, line 315 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22505#-L4qg-swWVvLSJGSa4fE:-L4reBqLNEaaZmcBtIND:bgtvdkn) ([raw file](https://github.com/cockroachdb/cockroach/blob/210d03c2432310f386fb2cfd6cfa079b8efbb6cc/pkg/sql/opt/build/builder.go#L315)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Whoops, didn't mean to change that. Fixed. Maybe we're missing test(s) to catch it?
</blockquote></details>

Yea I don't think many of the current tests are using AND/OR in SELECT expressions.  I can add some of those and any other expression types we're missing in another PR.

---

*[pkg/sql/opt/xform/factory.go, line 112 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22505#-L4qjKSRCRjqR-nz1HOm:-L4rf16Qe51o693sr_O0:b39mr3c) ([raw file](https://github.com/cockroachdb/cockroach/blob/210d03c2432310f386fb2cfd6cfa079b8efbb6cc/pkg/sql/opt/xform/factory.go#L112)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Because if we're flattening an And, there's going to be at least one more expression to add to the final list (and maybe more than one).
</blockquote></details>

Ah right - this is only called if there is at least one child AND node

---

*[pkg/sql/opt/xform/rules/bool.opt, line 51 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22505#-L4qn86HyCCeF56ABE9e:-L4rffZ_6z7zlfRoCj7T:b-e3hs27) ([raw file](https://github.com/cockroachdb/cockroach/blob/210d03c2432310f386fb2cfd6cfa079b8efbb6cc/pkg/sql/opt/xform/rules/bool.opt#L51)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

They're run in same order as file. In this case, it doesn't matter what order they're in, though. The rules trigger bottom-up, so `EliminateNot` will never trigger before `NegateComparison`.
</blockquote></details>

By "bottom up" you mean it matches the leaves of the expression tree first? I think I see what you mean.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22505)*
<!-- Sent from Reviewable.io -->

rytaft commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22505#-:-L4ri4KG7AzzeFSbPeQr:bnfp4nl)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt/xform/rules/bool.opt, line 51 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22505#-L4qn86HyCCeF56ABE9e:-L4skL47cSLpXrCdQv_O:bgcck37) ([raw file](https://github.com/cockroachdb/cockroach/blob/210d03c2432310f386fb2cfd6cfa079b8efbb6cc/pkg/sql/opt/xform/rules/bool.opt#L51)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

By "bottom up" you mean it matches the leaves of the expression tree first? I think I see what you mean.
</blockquote></details>

it means that if there's a NOT within a NOT, then the inner NOT no longer matches any pattern, because pattern matching occurs on it before its parent (i.e. bottom-up). And since the inner NOT doesn't match *any* pattern, then it must not match `NegateComparison` - it had first shot.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22505)*
<!-- Sent from Reviewable.io -->

rytaft commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt/xform/rules/bool.opt, line 51 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22505#-L4qn86HyCCeF56ABE9e:-L4w8X603RUi3CPVs6oF:b5vjmdl) ([raw file](https://github.com/cockroachdb/cockroach/blob/210d03c2432310f386fb2cfd6cfa079b8efbb6cc/pkg/sql/opt/xform/rules/bool.opt#L51)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

it means that if there's a NOT within a NOT, then the inner NOT no longer matches any pattern, because pattern matching occurs on it before its parent (i.e. bottom-up). And since the inner NOT doesn't match *any* pattern, then it must not match `NegateComparison` - it had first shot.
</blockquote></details>

Got it - thanks!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22505)*
<!-- Sent from Reviewable.io -->
